### DISCUSSION TITLE
Applying detekt-api in Android project fails with: ethodHandle.invoke and MethodHandle.invokeExact are only supported starting with Android O

### DISCUSSION THREAD

**Discussion by svyd** [detekt/detekt#7567] [General]
## Expected Behavior
Created a separate module in the existing Android project to add some custom rules for Detekt.
`implementation("io.gitlab.arturbosch.detekt:detekt-api:1.22.0")` doesn't allow to build project - 

> Failed to transform kotlin-compiler-embeddable-1.8.22.jar (org.jetbrains.kotlin:kotlin-compiler-embeddable:1.8.22) to match attributes {artifactType=android-global-synthetics, dexing-component-attributes=ComponentSpecificParameters(minSdkVersion=24, debuggable=true, enableCoreLibraryDesugaring=true, enableGlobalSynthetics=true, enableApiModeling=false, dependenciesClassesAreInstrumented=false, asmTransformComponent=null, useJacocoTransformInstrumentation=false, enableDesugaring=true, needsClasspath=false, useFullClasspath=false, componentIfUsingFullClasspath=null), org.gradle.category=library, org.gradle.libraryelements=jar, org.gradle.status=release, org.gradle.usage=java-runtime}.
      > Execution failed for DexingNoClasspathTransform: /Users/vsvydenko/.gradle/caches/transforms-4/cf2301bb46c1d86b1bcc2ed83a7c0f13/transformed/jetified-kotlin-compiler-embeddable-1.8.22.jar.
         > Error while dexing.
           Increase the minSdkVersion to 26 or above.

Caused by: [CIRCULAR REFERENCE: com.android.tools.r8.internal.h: MethodHandle.invoke and MethodHandle.invokeExact are only supported starting with Android O (--min-api 26): Lorg/jetbrains/kotlin/com/intellij/util/io/ByteBufferUtil;cleanBuffer(Ljava/nio/ByteBuffer;)Z]

If `implementation("io.gitlab.arturbosch.detekt:detekt-api:1.22.0")` comment this dependency - I can build project.

## Context
My project has minSdk 24. Changing to 26 doesn't fix problem.
detekt-custom-rules module:
build. gradle.kts
```
plugins {
    id("kotlin")
}

dependencies {
    implementation("org.jetbrains.kotlin:kotlin-stdlib:${Versions.Kotlin.SDK}")
    implementation("io.gitlab.arturbosch.detekt:detekt-api:1.22.0")
}
```

Root build.gradle:

```
plugins {
    id("io.gitlab.arturbosch.detekt").version("1.22.0")
}

dependencies {
    detektPlugins("io.gitlab.arturbosch.detekt:detekt-formatting:1.22.0")
    detektPlugins(project(":detekt-custom-rules"))
}

tasks.register("detektIssues", io.gitlab.arturbosch.detekt.Detekt::class.java) {
    buildUponDefaultConfig = true

    config.setFrom(files("${rootProject.projectDir}/detekt-custom-rules/src/main/resources/config/detekt.yml"))
    baseline.set(file("${rootProject.projectDir}/detekt-custom-rules/src/main/resources/config/detekt-baseline.xml"))

    description = "DETEKT build for all modules"
    parallel = true
    ignoreFailures = false
    autoCorrect = false
    setSource(file(rootProject.projectDir))
    include("**/*.kt")
    exclude("**/resources/**", "**/build/**")

    reports {
        xml.required.set(false)
        html.required.set(true)
        txt.required.set(true)
        sarif.required.set(false)
        md.required.set(false)
    }
}

val compileSdkVersion by extra { 34 }
val buildToolsVersion by extra { "34.0.0" }
val minSdkVersion by extra { 24 }
val targetSdkVersion by extra { 34 }

```

## Your Environment
```
./gradlew --version

------------------------------------------------------------
Gradle 8.7
------------------------------------------------------------

Build time:   2024-03-22 15:52:46 UTC
Revision:     650af14d7653aa949fce5e886e685efc9cf97c10

Kotlin:       1.9.22
Groovy:       3.0.17
Ant:          Apache Ant(TM) version 1.10.13 compiled on January 4 2023
JVM:          17.0.12 (Homebrew 17.0.12+0)
OS:           Mac OS X 14.5 aarch64

```

* Version of detekt used: 1.22.0
* Version of Gradle used (if applicable):
* Gradle scan link (add `--scan` option when running the gradle task):
* Operating System and version: provided above
* Link to your project (if it's a public repository): -

**Comment by 3flex** [comment]
Can you try with the newest detekt version? If that fails can you provide a build scan?

**Comment by svyd** [comment]
@3flex same issue with newest version 1.23.6

Tried with detekt version `1.23.6`

Same results.
Project sync - works.
Project build - fails
During build time, new Finder window displays with directory **transformed** and 
`jetified-kotlin-compiler-embeddable-1.9.23.jar` inside this folder.

Please check detekt-api-fails.txt attached with full output error.

As well, AS provide this messages:

```
2 files found with path 'kotlin/coroutines/coroutines.kotlin_builtins'.
Adding a packaging block may help, please refer to
https://developer.android.com/reference/tools/gradle-api/8.3/com/android/build/api/dsl/Packaging
for more information
```

BUild scan:[deleted]

Kotlin version in project: "2.0.0"

**Comment by 3flex** [comment]
We'll need a build scan that shows the failure in the log. 

It's also possible this just won't work if you're mixing android dependencies and Android Gradle plugin within a custom rules module.

**Comment by svyd** [comment]
@3flex 
Attached file 
./gradlew :Investing:build >> build-detekt-api-fails.txt
As well it throw error in terminal. attached as well - please see console_output_failed.txt

_DELETED LOGS_

Hope , this helps

**Comment by 3flex** [comment]
Run `./gradlew :Investing:build --scan`

**Comment by svyd** [comment]
@3flex please check: https://scans.gradle.com/s/lpkbordccen3s

**Comment by svyd** [comment]
**build.gradle.kts** for new module **detekt-custom-rules** - no android related dependencies

```
plugins {
    id("kotlin")
}

dependencies {
    implementation("org.jetbrains.kotlin:kotlin-stdlib:${Versions.Kotlin.SDK}")
    implementation("io.gitlab.arturbosch.detekt:detekt-api:${Versions.DETEKT}")
}
```

**Comment by 3flex** [comment]
I'm going to move this to a discussion as I'm not convinced there's a bug at this stage. Others might be able to jump in and help with your project configuration but I don't have capacity to help troubleshoot unfortunately.

**Comment by svyd** [comment]
If comment line `implementation("io.gitlab.arturbosch.detekt:detekt-api:${Versions.DETEKT}") `
```
plugins {
    id("kotlin")
}

dependencies {
    implementation("org.jetbrains.kotlin:kotlin-stdlib:${Versions.Kotlin.SDK}")
    //implementation("io.gitlab.arturbosch.detekt:detekt-api:${Versions.DETEKT}")
}
```

project build with no issues. So it looks like it is or wrong configuration or issues with **detekt-api** project..

**Comment by svyd** [comment]
Resolved.

To fix problem:
instead:
`implementation("io.gitlab.arturbosch.detekt:detekt-api:${Versions.DETEKT}")`
use:
`compileOnly("io.gitlab.arturbosch.detekt:detekt-api:${Versions.DETEKT}")`
